---
layout: default
overview: true
---

<script src="//code.jquery.com/jquery-3.2.1.min.js"></script>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.0.47/jquery.fancybox.min.css" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.0.47/jquery.fancybox.min.js"></script>


	
		<p><b>SimSearch</b> is a pattern-search algorithm that provides
			biologists with the ability, once they identify an interesting
			genomic pattern, to look for similar occurrences in the data, thus
			facilitating genomic data access and use. For example, such patterns
			can describe <b>gene expression regulatory DNA areas</b> including
			heterogeneous (epi)genomic features (e.g. <b>histone modification</b> and/or
			different <b>transcription factor binding regions</b>). It is possible to
			define complex patterns based on <b>perfect matches</b> in genome tracks
			(regions that must match), <b>partial matches</b> (regions that are allowed
			to be absent), and <b>negative matches</b> (for instance to search for
			regions distant from transcription start sites).</p>
			<a data-fancybox="gallery" href="images/figure1.png"><img style="width: 250px; vertical-align: center" src="images/figure1.png"></a>
		<p>The SimSearch algorithm has been developed for the <a href="http://gendata.weebly.com/">GenData
			project</a> as a collaboration between
			University of Bologna and Politecnico di Milano. The SimSearch App
			for IGB is developped by @arnaudceol for the Istituto Italiano di
			Tecnologia. It is now part of the <a href="http://www.bioinformatics.deib.polimi.it/geco/">DATA-DRIVEN GENOMIC COMPUTING
(GeCo) <img width="35px" src="http://www.bioinformatics.deib.polimi.it/geco/imgs/logo-geco.png"/></a> project.</p>
		<p>
			Project URL: <a
				href="http://www-db.disi.unibo.it/research/GenData/SimSearch ">http://www-db.disi.unibo.it/research/GenData/SimSearch</a>
				</p>
		
		
		
		<!-- ----------------------------------------------------------------------------------  -->
		
		<h2>Installation</h2>

		<h3>Install the Integrated Genome Browser (IGB)</h3>
		<p>The SimSearch App is a plugin for the Integrated Genome Browser (IGB)
		that can be downloaded from <a href="http://bioviz.org/igb/">http://bioviz.org/igb/</a>;</p>

		<h3>Add the GenData repository:</h3>
		<p>From IGB, go to the "<b>plug-ins</b>" tab and launch the app manager.</p> 
				
		<a data-fancybox="gallery"  href="images/app-manager.png"><img src="images/app-manager.png"/></a>
		
		<a data-fancybox="gallery" href="images/figure1.png"><img src="images/figure1.png"></a>
		
		<p>Add the GenData repository: press the "<b>add</b>" button, and insert
		"<b>SimSearch</b>" for name, and the url:
		<b>https://arnaudceol.github.io/simsearch/plugins/</b>. Then select the
		SimSearch plugin and press "<b>install</b>". A new tab will appear in the
		tab panel in IGB.
		</p>

		<h3>Add the SimSearch Quickload server</h3>
		<p>In order to make it easy to test the SimSearch app, we have compiled a
		Quickload repository with public data (mostly from ENCODE and Roadmap
		projects) for the human hg19 genome. </p> 
				
		<a data-fancybox="gallery"  href="images/quickload.png"><img src="images/quickload.png"/></a>
		
		<p>To access this repository go to
		"<b>data access</b>", "<b>configure</b>", "<b>add...</b>", and add the
		repository: name: "<b>SimSearch</b>" url:
		"<b>http://cru.genomics.iit.it/simsearch-quickload/</b>" Remember to
		select species "<b>Homo sapiens</b>", version "<b>H_sapiens_Feb_2009</b>" in
		order to see the datasets.</p>

		<!-- ----------------------------------------------------------------------------------  -->

		<h2>Video:</h2>

		<p>Have a look at <a href="https://www.youtube.com/watch?v=jW65ope1h1o">https://www.youtube.com/watch?v=jW65ope1h1o</a>
		to see the SimSearch app in action: </p><p>
	<iframe width="560" height="315" src="https://www.youtube.com/embed/jW65ope1h1o" frameborder="0" allowfullscreen></iframe>
		</p>

		<!-- ----------------------------------------------------------------------------------  -->
	
		<h2>Annotation of the results:</h2>
		<p>In order to to facilitate the analyses of the results, it is possible to annotate them automatically with different sets of information:</p>
		<table>
		<tr><th>Type</th><th>Description</th></tr>
		<tr><td>Closest gene</td><td>Gene whose TSS is closer to the center of the matching region of the first perfect match dataset, or to the TSS if it has been selected as one dataset of type "perfect match". </td></tr>
		<tr><td>Functional annotations</td><td>Use Pantherdb web service to see if the genes identified in the results (closest genes) are enriched in a particular pathway or biological process. </td></tr>
		<tr><td>Annotation from tracks</td><td>Use annotations from a track loaded in IGB and see if the region that overlap with the results are enriched in a particular annotation. For instance, it is possible to load the genome segmentation from ChromHMM or Segway to see if the results are associated to a particular chromatin state.</td></tr>
		<tr><td>Pattern combinations</td><td>When a search involved several partial match, retrieve the dataset that are found more often together.</td></tr>
		</table>
		
 
		
		<!-- ----------------------------------------------------------------------------------  -->
		
		<h2>Usage:</h2>
		<h3>Main panel</h3>
		<h4>Build panel</h4>
		
		<a data-fancybox="gallery"  href="images/menu.png"><img src="images/menu.png"/></a>
		
		<ul>
		<li><b>Selection:</b> </li>
		<li><b>Dataset type:</b> </li>
		<li><b>Align:</b> </li>
		<li><b>Motif:</b> </li>
		<li><b>TSS: </b> </li>
		<li><b>Clear:</b> </li>
		<li><b>All tracks:</b> </li>
		</ul>
				
				
		<h4>Misc.</h4>
		<ul>
		<li><b>Load pre-defined pattern:</b> </li>
		<li><b>Use attributes: </b> </li>
		<li><b>Advanced parameters: </b>
			<ul>
			<li><b>Minimum similarity score of results:</b> </li>
			<li><b>Minimum alignment score for perfect match:</b> </li>
			<li><b>Score for missing partial match:</b> score assigned when a partial dataset is missing.
			The lower the score, the more penalyzed will be the result for the absence of the dataset.</li>
			<li><b>Distance between regions based on: </b> used to decide whether to take into consideration 
			the distance between the centers of the region, or the distance between the right end of the 
			first region and the start of the second region. Normalized centroid distance is similar to the centroid distance except it divides it by half the length 
			of the region. It permits to avoid loosing information with large regions which may overlap although their centers are distant.
			</li>
			<li><b>distance mismatch computed using:</b> choose between L1 (Manhattan distance, uses the absolute value of the distance)
			 and L2 (Euclidian distance, uses the square of the distance).</li>
			<li><b>Alignment score sigmoid midpoint:</b> The sigmoid is the function used to calculate a similarity 
			score between 0 (big distance) and 1 (distance = 0). The midpoint sets for which distance the similarity is equals to 0.5.</li>
			<li><b>Aligment score sigmoid slope:</b>
			 </li>
			<li><b>Use diversity:</b> when calculating the similarity; it is possible to take in consideration the 
			attributes of the genomic regions (e.g. fold enrichment).</li>
			<li><b>Is region length relevant for computing region similarity:</b> If true, the matching regions should have similar length 
			to give a high score.</li>
			</ul>
		</li>
		</ul>
		
		<a data-fancybox="gallery"  href="images/igb-enhancer-searchtab.png"><img src="images/igb-enhancer-searchtab.png"/></a>
		
		
		<h4>Output</h4>
		<ul>
		<li><b>Label:</b> name given to the tab and the new track.</li>
		<li><b>Track color:</b> color for the new track created.</li>
		<li><b>Max results:</b> maximum number of results.</li>
		</ul>

		<h4>Submit</h4>
		<ul>
		<li><b>View pattern:</b> display a graphical representation of the pattern.</li>
		<li><b>Submit:</b> Submit the query.</li>
		</ul>
		
		<h3>Result panel</h3>
		<h4>Pattern summary</h4>
		<p>This panel shows a summary of each dataset selected and its type.
		A color is associated randomly to each dataset, the same color will be used in the result table. 
		</p>
		
		<h4>Result table</h4>
		<p></p>
		
		
		<a data-fancybox="gallery"  href="images/enhancer-results-table.png"><img src="images/enhancer-results-table.png"/></a>
		
		<ul>
		<li><b>Result index:</b> </li>
		<li><b>Score:</b> </li>
		<li><b>Position:</b> </li>
		<li><b>Length:</b> </li>
		<li><b>Closest gene:</b> </li>
		<li><b>Distance:</b> </li>
		<li><b>Up/down stream:</b> </li>
		<li><b>Similarity scores:</b> </li>
		<li><b>Alignment snapshot:</b> </li>
		</ul>
	
		<h4>Actions:</h4> 
		
		<ul>
		<li><b>Save:</b> </li>
		<li><b>Functional annotations:</b>
			<p></p>
			<ul>
			
			<li><b>Name:</b></li>
			<li><b>p-value:</b> </li>
			<li><b>Number of genes:</b> </li>
			<li><b>Gene list:</b> </li>
			</ul>
			
		</li>
		<li><b>Track annotations:</b> </li>
		<li><b>Summary:</b>
			<b></b> 
			<ul>
			<li><b>Datasets:</b> </li>
			<li><b>Pattern:</b> </li>
			</ul>
		</li>
		</ul>
	




 
 
 
		
		
