---
layout: default
overview: true
---

<script src="//code.jquery.com/jquery-3.2.1.min.js"></script>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.0.47/jquery.fancybox.min.css" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.0.47/jquery.fancybox.min.js"></script>

		<h2>Main panel</h2>
		<h3>Build panel</h3>
		
		<a data-fancybox="gallery"  href="images/menu.png"><img src="images/menu.png"/></a>
		
		<ul>
		<li><b>+ Selection:</b> Add selected elements to the pattern. It can be one or more regions
		selected in the browser, or one or more tracks (select the track name on the left of the browser).</li>
		<li><b>Dataset type:</b> Select wether the datasets added should be of type perfect match, partial math, negative match or 
		any other type. This can be changed later in the pattern table.</li>
		<li><b>Align:</b> Use to align all the selected datasets. This is useful is the user has selected
		the datasets from different parts of the browser, but want them to be found at th esame position.</li>
		<li><b>Motif:</b> Search for DNA motif, for instance a known transcription factor motif such as the e-box. 
		The motif should be entered as a string (only nucleotide letter: A, C, G ant T). Note that this will only work 
		if the DNA sequence has already been loaded in the browser.</li>
		<li><b>TSS: add the transcription start sites as a dataset.</b> </li>
		<li><b>Clear:</b> Remove all datasets from the pattern.</li>
		<li><b>All tracks:</b> Add all tracks as datasets, for instance if all the histone marks ChIP-seq tracks have been loaded.
		The datasets can be individually edited or removed later.</li>
		</ul>
				
				
		<h3>Misc.</h3>
		<ul>
		<li><b>Load pre-defined pattern:</b> load a predefined pattern (chromatine states inferred 
		from the results of ChromHMM and Segway). The app will try to associate each dataset to a track according to 
		both names. The user should verify that the mapping is correct.</li>
		<li><b>Use attributes:</b> use attributes of the target tracks when searching from the best match. 
		The attributes are selected individually for each track (see below: edit button).</li>
		<li><b>Advanced parameters: </b>
			<ul>
			<li><b>Minimum similarity score of results:</b> results with a lower score will be discarded</li>
			<li><b>Minimum alignment score for perfect match:</b> individual datasett alignments with a score below
			this threshold will be discarded. We usually select a low threshold (e.g. 0.01), to avoid matching regions 
			too distant ones from the others.</li>
			<li><b>Score for missing partial match:</b> Score assigned when a partial dataset is missing.
			The lower the score, the more penalyzed will be the result for the absence of the dataset.</li>
			<li><b>Distance between regions based on: </b> used to decide whether to take into consideration 
			the distance between the centers of the region, or the distance between the right end of the 
			first region and the start of the second region. Normalized centroid distance is similar to the centroid distance except it divides it by half the length 
			of the region. It permits to avoid loosing information with large regions which may overlap although their centers are distant.
			</li>
			<li><b>Distance mismatch computed using:</b> choose between L1 (Manhattan distance, uses the absolute value of the distance)
			 and L2 (Euclidian distance, uses the square of the distance).</li>
			<li><b>Alignment score sigmoid midpoint:</b> The sigmoid is the function used to calculate a similarity 
			score between 0 (big distance) and 1 (distance = 0). The midpoint sets for which distance the similarity is equals to 0.5.</li>
			<li><b>Aligment score sigmoid slope:</b>
			 </li>
			<li><b>Use diversity:</b> when calculating the similarity; it is possible to take in consideration the 
			attributes of the genomic regions (e.g. fold enrichment).</li>
			<li><b>Is region length relevant for computing region similarity:</b> If true, the matching regions should have similar length 
			to give a high score.</li>
			</ul>
		</li>
		</ul>
		
		<a data-fancybox="gallery"  href="images/igb-enhancer-searchtab.png"><img src="images/igb-enhancer-searchtab.png"/></a>
		
		
		<h3>Output</h3>
		<ul>
		<li><b>Label:</b> name given to the tab and the new track.</li>
		<li><b>Track color:</b> color for the new track created.</li>
		<li><b>Max results:</b> maximum number of results.</li>
		</ul>

		<h3>Submit</h3>
		<ul>
		<li><b>View pattern:</b> display a graphical representation of the pattern.</li>
		<li><b>Submit:</b> Submit the query.</li>
		</ul>
		
		<h2>Result panel</h2>
		<h3>Pattern summary</h3>
		<p>This panel shows a summary of each dataset selected and its type.
		A color is associated randomly to each dataset, the same color will be used in the result table. 
		</p>
		
		<h3>Result table</h3>
		<p></p>
		
		
		<a data-fancybox="gallery"  href="images/enhancer-results-table.png"><img src="images/enhancer-results-table.png"/></a>
		
		<ul>
		<li><b>Score:</b> Similarity score fo the match.</li>
		<li><b>Position:</b> Position of the match: chromosome: center of the lefter dataset - sequence of the righter dataset.</li>
		<li><b>Length:</b> distance between the centers of the lefter dataset and righter dataset.</li>
		<li><b>Closest gene:</b> Gene which TSS is closest to the center of the root dataset, i.e. the first perfect match. If the TSS is 
		itself a dataset, then its gene is selected.</li>
		<li><b>Distance:</b> Distance between the center of the root dataset and the TSS.</li>
		<li><b>Up/down stream:</b> Indicate if the match is up- or down-strem of the closest gene.</li>
		<li><b>Similarity scores:</b> Each dataset is represented by a colored square, whose height is 
		proportional to the score: a perfect match with a score of 1 is represented by a perfect square,
		 while a match with a score of 0.5 is half its height, and an empty space indicates that no
		 match was found.</li>
		<li><b>Alignment snapshot:</b> a graphical representation of the pattern matching is displayed. 
		The regions of the query pattern are displayed in red, the matching regions in blue. 
		This allows to an immediate understanding of which regions are present or missing,
		 and whether the regions are well aligned to the query pattern. </li>
		</ul>
	
		<h3>Actions:</h3> 
		
		<ul>
		<li><b>Save:</b> save the results in a tab delimited file.</li>
		<li><b>Functional annotations:</b>
			<p>Query <a href="http://pantherdb.org/">Pantherdb</a> to see if the genes 
			identified in the results (closest genes) are enriched in a 
			particular pathway or biological process.</p>
			<ul>
			
			<li><b>Name:</b> Name of the pathway or biological process.</li>
			<li><b>p-value:</b> P-value assigned to this enrichment.</li>
			<li><b>Number of genes:</b> Number of genes from the results
			associated to this term.</li>
			<li><b>Gene list:</b> List of genes in the results associated to this term.</li>
			</ul>
			
		</li>
		<li><b>Track annotations:</b> It is possible to select the annotations from one of the 
		tracks loaded in the browser (for instance chromatine states identified by ChrommHMM or Segway,
		available from the SimSearch quickload server). The app counts the annotations that overlap with the results,
		and perform a Fisher exact test to assess the significance of this association (the total number of annotations
		of the same type is used as background).
		
		</li>
		<li><b>Summary:</b>
			Show a summary of the results, i.e. the different "partial match" datasets which match together, and how many
			time they match together. Each individual combination can be selected to automatically
			create a new pattern that can be searched and analyzed. For a matter of clarity, smaller
			 combinations are only displayed if they are found more often that more complex ones (for instance, if 100 instances of A + B + C and 100 instances of A + B are found, only A+ B + C is displayed).
		</li>
		</ul>

		<h2>Annotation of the results:</h2>
		<p>In order to to facilitate the analyses of the results, it is possible to annotate them automatically with different sets of information:</p>
		<table>
		<tr><th>Type</th><th>Description</th></tr>
		<tr><td>Closest gene</td><td>Gene whose TSS is closer to the center of the matching region of the first perfect match dataset, or to the TSS if it has been selected as one dataset of type "perfect match". </td></tr>
		<tr><td>Functional annotations</td><td>Use Pantherdb web service to see if the genes identified in the results (closest genes) are enriched in a particular pathway or biological process. </td></tr>
		<tr><td>Annotation from tracks</td><td>Use annotations from a track loaded in IGB and see if the region that overlap with the results are enriched in a particular annotation. For instance, it is possible to load the genome segmentation from ChromHMM or Segway to see if the results are associated to a particular chromatin state.</td></tr>
		<tr><td>Pattern combinations</td><td>When a search involved several partial match, retrieve the dataset that are found more often together.</td></tr>
		</table>

		
	