

<!doctype html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Example: Distal enhancer &#8212; SimSearch App 1.5.1-igb9.1.2 documentation</title>
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.5.1-igb9.1.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Usage" href="usage.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">SimSearch App 1.5.1-igb9.1.2 documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><img src="_images/logo.PNG"/>
  <h4>Previous topic</h4>
  <p class="topless"><a href="usage.html"
                        title="previous chapter">Usage</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="example-distal-enhancer">
<h1>Example: Distal enhancer<a class="headerlink" href="#example-distal-enhancer" title="Permalink to this headline">¶</a></h1>
<p>In this example, we will look for distal enhancers (enhancers distant from a TSS), based on a pattern
inferred from the results of ChromHMM. We have already installed IGB, the SimSearch plugin and the SimSearch
Quickload server (see the installation page if you  have not done it yet).</p>
<div class="section" id="select-the-right-genome">
<h2>Select the right genome<a class="headerlink" href="#select-the-right-genome" title="Permalink to this headline">¶</a></h2>
<img alt="_images/ex1-genome.PNG" src="_images/ex1-genome.PNG" />
<dl class="docutils">
<dt>The data we will use has been aligned on the human genome build hg19, also called NCBI37, or Feb 2009. We select</dt>
<dd>the species Homo sapiens and the genome version H_sapiens_Feb_2019 in the right panel of IGB.</dd>
</dl>
</div>
<div class="section" id="load-tracks">
<h2>Load tracks<a class="headerlink" href="#load-tracks" title="Permalink to this headline">¶</a></h2>
<img alt="_images/ex1-tracks.PNG" src="_images/ex1-tracks.PNG" />
<p>We will load in new tracks the results of ChIP-seq experiments from ENCODE that are available
from the SimSearch Quickload server. We select this server in the data panel (bottom panel in IGB), and
select the Histone Marks available for the K562 cell line, as well as CTCF binding regions under tfbs.</p>
</div>
<div class="section" id="select-the-enhancer-pattern">
<h2>Select the enhancer pattern<a class="headerlink" href="#select-the-enhancer-pattern" title="Permalink to this headline">¶</a></h2>
<img alt="_images/ex1-select-predefined-pattern.PNG" src="_images/ex1-select-predefined-pattern.PNG" />
<p>Now we select an Enhancer pattern from the pre-defined pattern list. We choose the pattern named
Strong_Enhancer, which is inferred from the results of ChromHMM based on the data from ENCODE.</p>
</div>
<div class="section" id="add-tss">
<h2>Add TSS<a class="headerlink" href="#add-tss" title="Permalink to this headline">¶</a></h2>
<img alt="_images/ex1-tss.PNG" src="_images/ex1-tss.PNG" />
<p>Because we want a distal enhancer, we will add the TSS dataset, by pressing the TSS button in the “Build pattern” area of
the Search tab.</p>
<img alt="_images/ex1-rm-tss-edit.PNG" src="_images/ex1-rm-tss-edit.PNG" />
<p>We can edit this dataset: first we specify that we want it to be a negative dataset (we do not want the results
to contain a TSS), second we change the distance to 20000, to ensure that the results will be distant at least
20000 bases from a TSS.</p>
</div>
<div class="section" id="browsing-the-results">
<h2>Browsing the results<a class="headerlink" href="#browsing-the-results" title="Permalink to this headline">¶</a></h2>
<p>We are now ready to start the pattern search. The processing time depends on the number and complexity of the tracks,
as well as the internet connection (you can also use local files or a local Quickload repository).</p>
<a class="reference internal image-reference" href="_images/ex1-chromhmm-zoom.PNG"><img alt="_images/ex1-chromhmm-zoom.PNG" src="_images/ex1-chromhmm-zoom.PNG" style="width: 1024px;" /></a>
<p>The results are displayed in a table. If we click on any row, the plugin will automatically
zoom into the corresponding region in the browser, and display on the
right the regions associated with each dataset (blue) and how they aligned to the original pattern (red).</p>
</div>
<div class="section" id="analyzing-the-results">
<h2>Analyzing the results<a class="headerlink" href="#analyzing-the-results" title="Permalink to this headline">¶</a></h2>
<p>It is possible to browse the results, inspect them visually in the browser, and also run additional analyzes
directly from the plugin.</p>
<img alt="_images/ex1-chmmannots.PNG" src="_images/ex1-chmmannots.PNG" />
<p>We choose to compare our results to those of the original ChromHMM annotations. We load the results from ChromHMM
from the SimSearch Quickload server, in the annotation section of the K562 section. From the result panel, we select
this track and press the “Track annotations” button.</p>
<a class="reference internal image-reference" href="_images/ex1-chromhmm-annotations.PNG"><img alt="_images/ex1-chromhmm-annotations.PNG" src="_images/ex1-chromhmm-annotations.PNG" style="width: 540px;" /></a>
<p>The result table shows, as expected, an enrichment for enhancer regions.</p>
<a class="reference internal image-reference" href="_images/ex1-enhancer-vs-chromhmm-browse.PNG"><img alt="_images/ex1-enhancer-vs-chromhmm-browse.PNG" src="_images/ex1-enhancer-vs-chromhmm-browse.PNG" style="width: 1024px;" /></a>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">SimSearch App 1.5.1-igb9.1.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Arnaud Ceol.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.7.
    </div>
  </body>
</html>